### PR TITLE
Messaging

### PR DESCRIPTION
**Pull request by henninghaakonsen**

### CODE REVIEW

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/MessageConversationController.java [hunk 1]
```diff
@@ -266,7 +288,57 @@ public void postMessageConversationReply(
...
+
+
+        additionalUsers.forEach( user -> {
+            if ( !conversation.getUsers().contains( user ) )
+            {
+                conversation.addUserMessage( new UserMessage( user, false ) );
+            }
+        } );
+
+        conversation.getUsers().forEach( user -> System.out.println("user: " + user.getDisplayName()) );
```
Comment (stian-sandvold) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/MessageConversationController.java]: System.out.println

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/MessageConversationController.java [hunk 2]
```diff
@@ -414,7 +486,7 @@ public void postMessageConversationFeedback( @RequestParam( "subject" ) String s
             return responseNode;
         }
 
-        if ( !configurationService.isUserInFeedbackRecipientUserGroup( userToAssign ) )
+        if ( messageConversation.getMessageType() == MessageType.TICKET && !configurationService.isUserInFeedbackRecipientUserGroup( userToAssign ) )
```
Comment (stian-sandvold) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/MessageConversationController.java]: OR messageType is Validation result

Comment (henninghaakonsen) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/MessageConversationController.java]: Why? Isn't it just for tickets that we look at the feedback recipient group?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/message/hibernate/HibernateMessageConversationStore.java [hunk 3]
```diff
@@ -68,6 +67,23 @@ public void setStatementBuilder( StatementBuilder statementBuilder )
...
+    {
+        String userMessagesSql = "select distinct mc.messageconversationid " +
+            "from messageconversation mc INNER JOIN messageconversation_usermessages as mcu " +
+            "on mc.messageconversationid = mcu.messageconversationid " +
+            "INNER JOIN usermessage as um on mcu.usermessageid = um.usermessageid  where um.userid = " + user.getId();
+
+        SqlRowSet sqlRowSet = jdbcTemplate.queryForRowSet( userMessagesSql );
+
+        List<Integer> userMessages = new ArrayList<>( );
+        while (sqlRowSet.next()) {
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/message/hibernate/HibernateMessageConversationStore.java]: Codestyle

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/message/hibernate/HibernateMessageConversationStore.java [hunk 4]
```diff
@@ -68,6 +67,23 @@ public void setStatementBuilder( StatementBuilder statementBuilder )
     // Implementation methods
     // -------------------------------------------------------------------------
 
+    private List<Integer> getUserMessagesForUser( User user )
+    {
+        String userMessagesSql = "select distinct mc.messageconversationid " +
+            "from messageconversation mc INNER JOIN messageconversation_usermessages as mcu " +
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/message/hibernate/HibernateMessageConversationStore.java]: Inner join on new line

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/message/hibernate/HibernateMessageConversationStore.java [hunk 5]
```diff
@@ -68,6 +67,23 @@ public void setStatementBuilder( StatementBuilder statementBuilder )
...
+            "on mc.messageconversationid = mcu.messageconversationid " +
+            "INNER JOIN usermessage as um on mcu.usermessageid = um.usermessageid  where um.userid = " + user.getId();
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/message/hibernate/HibernateMessageConversationStore.java]: Lowercase SQL

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/MessageConversationController.java [hunk 6]
```diff
@@ -266,13 +288,78 @@ public void postMessageConversationReply(
...
 
+    //
+
+
+    @RequestMapping( value = "/{uid}/recipients", method = RequestMethod.POST )
+    public void addRecipientsToMessageConversation( @PathVariable( "uid" ) String uid, @RequestBody MessageConversation messageConversation,
+        HttpServletRequest request, HttpServletResponse response )
+        throws Exception
+    {
+        org.hisp.dhis.message.MessageConversation conversation = messageService.getMessageConversation( uid );
```
Comment (larshelge) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/MessageConversationController.java]: This seems to be a bit too much code for the presentation layer.

**Pull request merged**